driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -686,6 +690,12 @@ public void messageReceived(ChannelHandlerContext ctx, MessageEvent e) {
             }
         }
 
+        @Override
+        public void channelIdle(ChannelHandlerContext ctx, IdleStateEvent e) throws Exception {
```

Comment:
I've noticed that it's possible for a heartbeat to be sent while one is outstanding if readTimeout is greater than heartbeat interval (which doesn't seem like a very likely usecase).  I don't think this is a concern as if the first heartbeat times out the connection will ultimately be closed anyways.  We could add logic to make sure only one heartbeat is in progress for a connection at a time, but that might not be worth the added complexity, what do you think?